File: src/main/1.3/TwoSumTimed/TwoSumTimed.py
"""
Benchmark version of TwoSum - measures only algorithm performance
No user interaction - runs automated timing tests
"""

import time

def find_two_sum(numbers, target):
    """Find two numbers that sum to target"""
    n = len(numbers)
    for i in range(n - 1):
        for j in range(i + 1, n):
            if numbers[i] + numbers[j] == target:
                return (numbers[i], numbers[j], i, j)
    return None

def run_single_test():
    """Run single execution with timing"""
    test_array = [10, 15, 3, 7, 22, 8, 12, 5]
    target = 18
    
    print("\n=== Single Execution Test ===")
    print(f"Array: {test_array}")
    print(f"Target: {target}\n")
    
    start_time = time.perf_counter()
    result = find_two_sum(test_array, target)
    end_time = time.perf_counter()
    
    execution_time = end_time - start_time
    time_microseconds = execution_time * 1000000
    
    if result:
        num1, num2, idx1, idx2 = result
        print(f"Result: Found {num1} + {num2} = {target}")
    else:
        print("Result: No pair found")
    
    print(f"Execution time: {execution_time:.9f} seconds ({time_microseconds:.3f} microseconds)")

def run_benchmark():
    """Run multiple iterations for accurate timing"""
    test_array = [10, 15, 3, 7, 22, 8, 12, 5]
    target = 18
    iterations = 100000
    
    print("\n=== Benchmark Test (Multiple Iterations) ===")
    print(f"Running {iterations} iterations...")
    
    start_time = time.perf_counter()
    
    for _ in range(iterations):
        find_two_sum(test_array, target)
    
    end_time = time.perf_counter()
    
    total_time = end_time - start_time
    avg_time = total_time / iterations
    avg_microseconds = avg_time * 1000000
    
    print(f"Total time: {total_time:.6f} seconds")
    print(f"Average per iteration: {avg_time:.9f} seconds ({avg_microseconds:.3f} microseconds)")
    print(f"Iterations per second: {iterations / total_time:.0f}\n")

def main():
    print("\n╔════════════════════════════════════════╗")
    print("║  Two Sum - Python Performance Benchmark║")
    print("╚════════════════════════════════════════╝")
    
    run_single_test()
    run_benchmark()

if __name__ == "__main__":
    main()

"""
Execution: python3 TwoSum_benchmark.py
"""

File: src/main/1.3/TwoSumTimed/TwoSumTimed.c
/*
 * Benchmark version of TwoSum - measures only algorithm performance
 * No user interaction - runs automated timing tests
 */

#include <stdio.h>
#include <stdbool.h>
#include <time.h>

typedef struct {
    bool found;
    int firstNum;
    int secondNum;
    int firstIndex;
    int secondIndex;
} TwoSumResult;

TwoSumResult findTwoSum(int *numbers, int size, int target) {
    TwoSumResult result = {false, 0, 0, -1, -1};
    
    for (int i = 0; i < size - 1; i++) {
        for (int j = i + 1; j < size; j++) {
            if (numbers[i] + numbers[j] == target) {
                result.found = true;
                result.firstNum = numbers[i];
                result.secondNum = numbers[j];
                result.firstIndex = i;
                result.secondIndex = j;
                return result;
            }
        }
    }
    return result;
}

void runSingleTest() {
    int testArray[] = {10, 15, 3, 7, 22, 8, 12, 5};
    int arraySize = 8;
    int target = 18;
    
    printf("\n=== Single Execution Test ===\n");
    printf("Array: [10, 15, 3, 7, 22, 8, 12, 5]\n");
    printf("Target: %d\n\n", target);
    
    clock_t start = clock();
    TwoSumResult result = findTwoSum(testArray, arraySize, target);
    clock_t end = clock();
    
    double time_seconds = ((double)(end - start)) / CLOCKS_PER_SEC;
    double time_microseconds = time_seconds * 1000000;
    
    if (result.found) {
        printf("Result: Found %d + %d = %d\n", 
               result.firstNum, result.secondNum, target);
    } else {
        printf("Result: No pair found\n");
    }
    
    printf("Execution time: %.9f seconds (%.3f microseconds)\n", 
           time_seconds, time_microseconds);
}

void runBenchmark() {
    int testArray[] = {10, 15, 3, 7, 22, 8, 12, 5};
    int arraySize = 8;
    int target = 18;
    int iterations = 100000;
    
    printf("\n=== Benchmark Test (Multiple Iterations) ===\n");
    printf("Running %d iterations...\n", iterations);
    
    clock_t start = clock();
    
    for (int i = 0; i < iterations; i++) {
        findTwoSum(testArray, arraySize, target);
    }
    
    clock_t end = clock();
    
    double total_time = ((double)(end - start)) / CLOCKS_PER_SEC;
    double avg_time = total_time / iterations;
    double avg_microseconds = avg_time * 1000000;
    
    printf("Total time: %.6f seconds\n", total_time);
    printf("Average per iteration: %.9f seconds (%.3f microseconds)\n", 
           avg_time, avg_microseconds);
    printf("Iterations per second: %.0f\n\n", iterations / total_time);
}

int main() {
    printf("\n╔════════════════════════════════════════╗\n");
    printf("║  Two Sum - C Performance Benchmark    ║\n");
    printf("╚════════════════════════════════════════╝\n");
    
    runSingleTest();
    runBenchmark();
    
    return 0;
}

/*
 * Compilation: gcc -O2 -o TwoSum_benchmark TwoSum_benchmark.c
 * Execution: ./TwoSum_benchmark
 */